plugins/modules/s3_object_info.py

```diff
@@ -0,0 +1,806 @@
+            all_facts[key] = {}
+            all_facts[key] = describe_s3_object_lock_configuration(connection, bucket_name)
+        elif key == 'object_retention':
+            all_facts[key] = {}
+            all_facts[key] = describe_s3_object_retention(connection, bucket_name, object_name)
+        elif key == 'object_tagging':
+            all_facts[key] = {}
+            all_facts[key] = describe_s3_object_tagging(connection, bucket_name, object_name)
+
+    return all_facts
+
+
+def get_object(connection, bucket_name, object_name):
+    params = {}
+    params['Bucket'] = bucket_name
+    params['Key'] = object_name
+
+    result = {}
+    object_info = {}
+
+    try:
+        object_info = connection.head_object(**params)
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+        pass
+
+    if len(object_info) != 0:
+        # Remove ResponseMetadata from object_info, convert to snake_case
+        del(object_info['ResponseMetadata'])
+        object_info = camel_dict_to_snake_dict(object_info)
+
+    result['object_data'] = object_info
+
+    return result
+
+
+@AWSRetry.jittered_backoff(retries=10)
+def _list_bucket_objects(connection, **params):
+    paginator = connection.get_paginator('list_objects')
+    return paginator.paginate(**params).build_full_result()
+
+
+def list_bucket_objects(connection, module, bucket_name):
+    params = {}
+    params['Bucket'] = bucket_name
+
+    result = []
+    list_objects_response = {}
+
+    try:
+        list_objects_response = _list_bucket_objects(connection, **params)
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+        module.fail_json_aws(e, msg='Failed to list bucket objects.')
+
+    if len(list_objects_response) != 0:
+        # convert to snake_case
+        for response_list_item in list_objects_response['Contents']:
+            result.append(response_list_item['Key'])
+
+    return result
+
+
+def bucket_check(connection, module, bucket_name,):
+    try:
+        connection.head_bucket(Bucket=bucket_name)
+    except is_boto3_error_code(['404', '403']) as e:
+        module.fail_json_aws(e, msg="The bucket %s does not exist or is missing access permissions." % bucket_name)
+
+
+def object_check(connection, module, bucket_name, object_name):
+    try:
+        connection.head_object(Bucket=bucket_name, Key=object_name)
+    except is_boto3_error_code(['404', '403']) as e:
+        module.fail_json_aws(e, msg="The object %s does not exist or is missing access permissions." % object_name)
+
+
+# To get S3 connection, in case of dealing with ceph, dualstack, etc.
+def is_fakes3(endpoint_url):
+    """ Return True if endpoint_url has scheme fakes3:// """
+    if endpoint_url is not None:
+        return urlparse(endpoint_url).scheme in ('fakes3', 'fakes3s')
+    else:
+        return False
+
+
+def get_s3_connection(module, aws_connect_kwargs, location, ceph, endpoint_url, sig_4=False):
```

Comment:
We can keep it as it is for the moment. For the future, we should move this in module_utils instead of introducing redundancy and refactor the other s3 modules as needed.

Reply:
Agreed. Makes sense.
Once this (and other s3 PRs if any) are merged, I can work on refactoring this.